### PR TITLE
Refactor field of law repository and service by using criteria query api

### PR DESCRIPTION
**Pull request by elaydis**
[RISDEV-5661](https://digitalservicebund.atlassian.net/browse/RISDEV-5661)

[RISDEV-5661]: https://digitalservicebund.atlassian.net/browse/RISDEV-5661?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: backend/src/main/java/de/bund/digitalservice/ris/caselaw/adapter/FieldOfLawService.java [hunk 1]
```diff
@@ -57,71 +56,34 @@ public Slice<FieldOfLaw> getFieldsOfLawBySearchQuery(
...
-        // Returned list is already ordered by identifier, so scoring is not necessary here
-        unorderedList = repository.findByIdentifier(identifierStr, pageable);
-        return sliceResults(unorderedList, pageable);
-      } else {
-        unorderedList = getResultsWithIdentifier(identifierStr, searchTerms, normStr);
-      }
-    } else {
-      unorderedList = getResultsWithoutIdentifier(searchTerms, normStr);
-    }
+    unorderedList =
```
Comment (kiwikern) [inline backend/src/main/java/de/bund/digitalservice/ris/caselaw/adapter/FieldOfLawService.java:71]: nitpick: can be combined

File: backend/src/main/java/de/bund/digitalservice/ris/caselaw/adapter/database/jpa/PostgresFieldOfLawRepositoryImpl.java [hunk 2]
```diff
@@ -72,119 +76,57 @@ public Slice<FieldOfLaw> findAllByOrderByIdentifierAsc(Pageable pageable) {
...
   @Transactional
-  public List<FieldOfLaw> findBySearchTerms(String[] searchTerms) {
-    if (searchTerms == null || searchTerms.length == 0) {
-      return Collections.emptyList();
+  public List<FieldOfLaw> findByCombinedCriteria(
+      String identifier, String[] descriptionSearchTerms, String[] normSearchTerms) {
+    CriteriaBuilder cb = entityManager.getCriteriaBuilder();
+    CriteriaQuery<FieldOfLawDTO> cq = cb.createQuery(FieldOfLawDTO.class);
+    Root<FieldOfLawDTO> fieldOfLawRoot = cq.from(FieldOfLawDTO.class);
+    fieldOfLawRoot.fetch("norms", JoinType.LEFT);
```
Comment (kiwikern) [inline backend/src/main/java/de/bund/digitalservice/ris/caselaw/adapter/database/jpa/PostgresFieldOfLawRepositoryImpl.java:84]: question: Do we always need to fetch with `norms` or only when we query by it? (There is also a checkbox "Mit Normen", not sure if that is related.)
But even if we don't always need it it would just be a performance optimization, so no blocker.

File: backend/src/main/java/de/bund/digitalservice/ris/caselaw/adapter/database/jpa/PostgresFieldOfLawRepositoryImpl.java [hunk 3]
```diff
@@ -72,119 +76,57 @@ public Slice<FieldOfLaw> findAllByOrderByIdentifierAsc(Pageable pageable) {
...
+    ArrayList<Predicate> predicates = new ArrayList<>();
+
+    Predicate notationPredicate =
+        cb.equal(fieldOfLawRoot.get("notation"), cb.literal(Notation.NEW));
```
Comment (kiwikern) [inline backend/src/main/java/de/bund/digitalservice/ris/caselaw/adapter/database/jpa/PostgresFieldOfLawRepositoryImpl.java:88]: question: Do we ever search by `Notation.NEW` or do we have them just for legacy doc units?

File: backend/src/main/java/de/bund/digitalservice/ris/caselaw/adapter/database/jpa/PostgresFieldOfLawRepositoryImpl.java [hunk 4]
```diff
@@ -72,119 +76,57 @@ public Slice<FieldOfLaw> findAllByOrderByIdentifierAsc(Pageable pageable) {
...
-        repository.findAllByNotationAndIdentifierContainingIgnoreCaseOrTextContainingIgnoreCase(
-            searchTerms[0]);
-
-    if (searchTerms.length == 1) {
-      return listWithFirstSearchTerm.stream()
-          .map(PostgresFieldOfLawRepositoryImpl::getWithNormsWithoutChildren)
-          .toList();
+    if (identifier != null) {
+      Predicate identifierPredicate =
+          cb.like(fieldOfLawRoot.get("identifier"), identifier.toUpperCase() + "%");
```
Comment (kiwikern) [inline backend/src/main/java/de/bund/digitalservice/ris/caselaw/adapter/database/jpa/PostgresFieldOfLawRepositoryImpl.java:99]: question: For the other search params, we search by `contains` and only for this one with `startsWith`. Is that correct? (I assume the identifier is known, so it does make sense to me, just wanted to double check.)

File: backend/src/test/java/de/bund/digitalservice/ris/caselaw/integration/tests/FieldOfLawIntegrationTest.java [hunk 5]
```diff
@@ -186,7 +186,11 @@ void testGetFieldsOfLawByNormsQuery_withoutAbbreviation(String query) {
   @ParameterizedTest
   @ValueSource(
       strings = {
+        "abc § 123", // norm followed by paragraph
+        "abc §123", // norm followed by paragraph without whitespace
+        "abc § 12", // norm followed by incomplete paragraph
         "§ 123 abc", // paragraph followed by norm
+        "§ 12 abc", // incomplete paragraph followed by norm
```
Comment (kiwikern) [inline backend/src/test/java/de/bund/digitalservice/ris/caselaw/integration/tests/FieldOfLawIntegrationTest.java:193]: praise: Nice to cover many edge cases. 👏  In general, these might hint that a unit test might be useful for the preprocessing, but an integration test on the database also works. (It does not start a new app for each parametrized test right?)

**Review by kiwikern — APPROVED** [review]
Amazing how much you have reduced the code, so much removed code. 🤩  Wow! Thank you for this really great refactoring, I love it! 😍 
Found no blockers at all, so feel free to merge. :-) All comments are optional.

File: backend/src/main/java/de/bund/digitalservice/ris/caselaw/adapter/database/jpa/PostgresFieldOfLawRepositoryImpl.java (hunk 3 above)
Comment (elaydis) [reply backend/src/main/java/de/bund/digitalservice/ris/caselaw/adapter/database/jpa/PostgresFieldOfLawRepositoryImpl.java:88]: I assume you mean `Notation.OLD`, and that's exactly right, those are deprecated and just exist in the table for the old doc units :)

File: backend/src/main/java/de/bund/digitalservice/ris/caselaw/adapter/database/jpa/PostgresFieldOfLawRepositoryImpl.java (hunk 4 above)
Comment (elaydis) [reply backend/src/main/java/de/bund/digitalservice/ris/caselaw/adapter/database/jpa/PostgresFieldOfLawRepositoryImpl.java:99]: yup! we always assume they know at least the first two letters of the identifier they want to search in, so it made sense to me that way :)
